pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java

```diff
@@ -79,29 +77,12 @@ private void attachTag(FlowNode currentNode, String syntheticContext) {
         }
     }
 
-    public static class GraphListenerAction extends InvisibleAction implements RunAction2 {
+    @Extension
+    public static class FlowExecutionListenerImpl extends FlowExecutionListener {
         @Override
-        public void onLoad(Run<?, ?> r) {
-            if (r != null && r instanceof WorkflowRun) {
-                WorkflowRun run = (WorkflowRun) r;
-                attachListener(run);
-            }
-        }
-
-        @Override
-        public void onAttached(Run<?, ?> r) {
-            if (r != null && r instanceof WorkflowRun) {
-                WorkflowRun run = (WorkflowRun) r;
-                attachListener(run);
-            }
-        }
-
-        private void attachListener(WorkflowRun run) {
-            if (run != null) {
-                FlowExecution exec = run.getExecution();
-                if (exec != null && !exec.isComplete()) {
-                    exec.addListener(new SyntheticStageGraphListener());
-                }
+        public void onRunning(FlowExecution execution) {
+            if (execution != null && !execution.isComplete()) {
+                execution.addListener(new SyntheticStageGraphListener());
```

Comment:
Ok, tweaked to only add the `SyntheticStageGraphListener` if the corresponding `WorkflowRun` has the `ExecutionModelAction` on it - i.e., we know it's a Declarative Pipeline.